core/src/avm2/parameters.rs

```diff
@@ -0,0 +1,176 @@
+    ///
+    /// If the value is null or is undefined, 0.0 is returned.
+    /// If the object cannot be coerced to an f64, a TypeError 1050 is raised.
+    fn get_f64(
+        &self,
+        activation: &mut Activation<'_, 'gc>,
+        index: usize,
+    ) -> Result<f64, Error<'gc>>;
+
+    /// Gets the value at the given index and coerces it to a u32.
+    ///
+    /// If the value is null or is undefined, 0 is returned.
+    /// If the object cannot be coerced to a u32, a TypeError 1050 is raised.
+    fn get_u32(
+        &self,
+        activation: &mut Activation<'_, 'gc>,
+        index: usize,
+    ) -> Result<u32, Error<'gc>>;
+
+    /// Gets the value at the given index and coerces it to a i32.
+    ///
+    /// If the value is null or is undefined, 0 is returned.
+    /// If the object cannot be coerced to an i32, a TypeError 1050 is raised.
+    fn get_i32(
+        &self,
+        activation: &mut Activation<'_, 'gc>,
+        index: usize,
+    ) -> Result<i32, Error<'gc>>;
+
+    /// Gets the value at the given index and coerces it to a bool.
+    ///
+    /// If the value is null or is undefined, false is returned.
+    fn get_bool(&self, index: usize) -> bool;
+
+    /// Gets the value at the given index and coerces it to an AvmString.
+    ///
+    /// If the value is undefined, "undefined" is returned.
+    /// If the value is null, "null" is returned.
+    /// If the object cannot be coerced to a string, a TypeError 1050 is raised.
+    fn get_string(
+        &self,
+        activation: &mut Activation<'_, 'gc>,
+        index: usize,
+    ) -> Result<AvmString<'gc>, Error<'gc>>;
+
+    /// Gets the value at the given index and coerces it to an AvmString.
+    ///
+    /// If the value is null or is undefined, Ok(None) is returned.
+    /// If the object cannot be coerced to a string, a TypeError 1050 is raised.
+    fn try_get_string(
+        &self,
+        activation: &mut Activation<'_, 'gc>,
+        index: usize,
+    ) -> Result<Option<AvmString<'gc>>, Error<'gc>>;
+}
+
+impl<'gc> ParametersExt<'gc> for &[Value<'gc>] {
+    fn get_object(
+        &self,
+        activation: &mut Activation<'_, 'gc>,
+        index: usize,
+        name: &'static str,
+    ) -> Result<Object<'gc>, Error<'gc>> {
+        match self[index] {
+            Value::Null | Value::Undefined => Err(null_parameter_error(activation, name)),
+            Value::Object(o) => Ok(o),
+            primitive => Ok(PrimitiveObject::from_primitive(primitive, activation)
+                .expect("Primitive object is infallible at this point")),
+        }
+    }
+
+    fn try_get_object(
+        &self,
+        activation: &mut Activation<'_, 'gc>,
+        index: usize,
+    ) -> Option<Object<'gc>> {
+        match self[index] {
+            Value::Null | Value::Undefined => None,
+            Value::Object(o) => Some(o),
+            primitive => Some(
+                PrimitiveObject::from_primitive(primitive, activation)
```

Comment:
in most places we don't actually expect a primitive (matrix, color_transform, etc), so I'd consider making it a default? (just documenting it well).

On that note, I'd also add `get_value(idx)` just for a consistent api.

Reply:
Added get_value which is `this[index]` for now.
Good point about non-primitives, I think I'll need to consider that a bit though. Will handle it in a future PR if that's ok!